File: mongodb-service/src/models/aggregationModel.ts
import * as mongoose from "mongoose";
import { Model } from "mongoose";

/**
 * Aggregation type, model mongoose document and schema definition. Models are used by
 * Controllers so they are picked up by the swagger. Documents and schemas are used by 
 * Repositories so model defined Mongodb functions can be used.
 */
type AggregationType = IAggregation & mongoose.Document;
export interface IAggregationModel {
    _id?: string;
    operations: Array<string>;
    featureColumns: Array<string>;
    jobId: string;
    metricColumn: string;
    name: string;
    sortColumnName: string;
}

export interface IAggregation extends IAggregationModel, mongoose.Document {
    _id: string;
}

export const AggregationSchema = new mongoose.Schema({
    operations: {
        required: true,
        type: Array
    },
    featureColumns: {
        required: true,
        type: Array
    },
    jobId: {
        required: true,
        type: String
    },
    metricColumn: {
        required: true,
        type: String
    },
    name: {
        required: true,
        type: String
    },
    sortColumnName: {
        required: true,
        type: String
    }
});

const Aggregation: Model<IAggregation> = mongoose.model<AggregationType>("Aggregation", AggregationSchema);
export default Aggregation;

File: analytic-engine/src/main/java/com/mycompany/models/AggregationModel.java
package com.mycompany.models;

import java.util.List;
import java.util.Objects;

/**
 * Represents an Aggregation object. Corresponds to the IAggregationModel type in the mongodb-service.
 */
public class AggregationModel {
    public String _id;
    public String __v;
    public List<OperationEnum> operations;
    public List<String> featureColumns;
    public String jobId;
    public String metricColumn;
    public String name;
    public String sortColumnName;

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        AggregationModel that = (AggregationModel) o;
        return Objects.equals(_id, that._id) &&
                Objects.equals(__v, that.__v) &&
                Objects.equals(operations, that.operations) &&
                Objects.equals(featureColumns, that.featureColumns) &&
                Objects.equals(jobId, that.jobId) &&
                Objects.equals(metricColumn, that.metricColumn) &&
                Objects.equals(name, that.name) &&
                Objects.equals(sortColumnName, that.sortColumnName);
    }
}
